src/compiler/checker.ts

```diff
@@ -2788,10 +2788,26 @@ namespace ts {
             }
         }
 
-        function addInheritedSignatures(signatures: Signature[], baseSignatures: Signature[]) {
-            if (baseSignatures) {
-                for (let signature of baseSignatures) {
-                    signatures.push(signature);
+        function addInheritedSignatures(baseType: ObjectType, filter: ResolveFilter) {
+            if (!(<ResolvedType>baseType).members) {
+                resolveClassOrInterfaceMembers(<InterfaceType>baseType, filter);
+            }
+            addFiltered(filter.callSignatures, (<ResolvedType>baseType).callSignatures, filter);
+            addFiltered(filter.constructSignatures, (<ResolvedType>baseType).constructSignatures, filter);
+            for (let symbolId in filter.excludeLater) {
+                filter.excludeInherited[symbolId] = true;
+            }
+            filter.excludeLater = {};

```

Comment:
Do you need to postpone the exclusion until the next iteration? What happens if you update the exclude bag as you go without even having excludeLater?